src/util/Exception.h

```diff
@@ -81,16 +84,54 @@ class Exception : public std::exception {
+// etc. Return "condition; <concatentation of `getMessageImpl(messages)...`>
+std::string concatMessages(std::string_view condition, auto&&... messages) {
+  if constexpr (sizeof...(messages) == 0) {
+    return std::string{condition};
+  } else {
+    return absl::StrCat(getMessageImpl(condition), "; ",
+                        getMessageImpl(messages)...);
+  }
+}
+}  // namespace ad_utility::detail
+
 // Custom assert which does not abort but throws an exception. Use this for
 // conditions that can be violated by calling a public (member) function with
 // invalid inputs. Since it is a macro, the code coverage check will only report
 // a partial coverage for this macro if the condition is never violated, so make
 // sure to integrate a unit test that violates the condition.
-#define AD_CONTRACT_CHECK(condition)                                                                                                                       \
+// The first argument is the condition (anything that is convertible to `bool`).
+// Additional arguments are concatenated to get a detailed error message in the
+// failure case. each of these arguments can be either a type that can be
+// converted to a string via `absl::StrCat` (e.g. strings or builtin numeric
+// types) or a callable that produce a `std::string`. The latter case is useful
+// if the error message is expensive to construct because the callables are only
+// invoked if the assertion fails. For examples see `ExceptionTest.cpp`.
+#define AD_CONTRACT_CHECK(condition, ...)                                                                                                                  \
   if (!(condition)) [[unlikely]] {                                                                                                                         \
     using namespace std::string_literals;                                                                                                                  \
     AD_THROW(                                                                                                                                              \
-        "Assertion `"s + std::string(__STRING(condition)) +                                                                                                \
+        "Assertion `"s +                                                                                                                                   \
+        ad_utility::detail::concatMessages(std::string(__STRING(condition))                                                                                \
+                                               __VA_OPT__(, ) __VA_ARGS__) +                                                                               \
         "` failed. Likely cause: A function was called with arguments that violate the contract of the function. Please report this to the developers."s); \
```

Comment:
1. The failed should come right after the condition, should it not?
2. Shouldn't the "Likely cause: ..." part only be there if there are no further arguments (with an explicit message)?